### ISSUE TITLE
[no-unnecessary-condition] Conflicts with TS strict type checking on object fields

### ISSUE THREAD

**Issue by mdouglass** [typescript-eslint/typescript-eslint#1200]
**Repro**

```JSON
{
  "parser": "@typescript-eslint/parser",
  "parserOptions": { "project": "./tsconfig.json", "sourceType": "module" },
  "plugins": ["@typescript-eslint"],
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": ["error", { "ignoreRhs": true }]
  }
}
```

```TS
interface ICount { count: number }
const field: Record<string, ICount> = {}
for (const key of ['matt','not-matt','matt']){
  field[key] = field[key] ?? { count: 0 }
  field[key].count = field[key].count + 1
}

```

**Expected Result**
No errors

**Actual Result**
```
  4:16  error  Unnecessary conditional, value is always truthy  @typescript-eslint/no-unnecessary-condition
```

**Additional Info**
So, I kind of understand what is happening here because field is defined as a `Record<string, ICount>` so it assumes that `field[key]` is `ICount` and therefore can't be `undefined`.

I tried the alternative (arguably, more accurate definition) of `Record<string, ICount|undefined>` but that runs headlong into TypeScript compiler errors:

```TS
interface ICount { count: number }
const field: Record<string, ICount | undefined> = {}
for (const key of ['matt','not-matt','matt']){
  field[key] = field[key] ?? { count: 0 }
  field[key].count = field[key].count + 1
}
```

```
lib/test-lint.ts:5:3 - error TS2532: Object is possibly 'undefined'.
5   field[key].count = field[key].count + 1
    ~~~~~~~~~~

lib/test-lint.ts:5:22 - error TS2532: Object is possibly 'undefined'.
5   field[key].count = field[key].count + 1
                       ~~~~~~~~~~
  ```

**Versions**

| package | version |
| ---------------------------------- | ------- |
| @typescript-eslint/eslint-plugin | 2.6.1 |
| @typescript-eslint/parser | 2.6.1 |
| TypeScript | 3.7.2 |
| ESLint | 6.6.0 |
| node | 13.1.0 |
| npm | 6.12.1 |

**Comment by bradzacher** [comment]
It's a sucky situation, and there's no way around it, unfortunately.

The problem is that typescript has no concept of object properties on records being potentially missing, as you noted.
The rule cannot account for it, because it only knows the types given by typescript, and typescript tells us the value must be an `ICount`, which ofc triggers the rule.

---

In your second example, typescript doesn't do a great job of refining the types; because you're using dynamic property access.
For example, if you were using static property access, typescript doesn't actually error:
```ts
interface ICount { count: number }
const field: Record<string, ICount | undefined> = {}

field.matt = field.matt ?? { count: 0 }
field.matt.count = field.matt.count + 1
// no errors, because typescript refines the type of `field.matt` to be not-undefined
```

You work get around this whilst using dynamic property access.
You have to assign the value to an intermediate variable, which typescript can knowingly refine the type of:
```ts
interface ICount { count: number }
const field: Record<string, ICount | undefined> = {}
for (const key of ['matt','not-matt','matt']){
  const val = field[key] = field[key] ?? { count: 0 }
  val.count = val.count + 1
}
```

---

Closing this as there's nothing we can do from the rule's POV.
